feat: 상세 화면 UI 구성 및 레이아웃 설계

Benjamin8282 opened the pull request:
이 Pull Request는 날씨 애플리케이션의 기능 확장, UI 개선, 백엔드 데이터 처리 강화를 포함한 주요 업데이트를 도입합니다. 핵심 변경 사항은 날씨 상세 화면의 추가, 새로운 데이터 모델 도입, 날씨 아이콘용 드로어블 리소스 추가입니다. 아래는 주요 변경 사항에 대한 세부 설명입니다:

### 기능 향상
* **날씨 상세 화면**: `WeatherDetailScreen`이 새롭게 디자인되어, 온도, 습도, 풍속, 시간대별 예보 등 상세 날씨 정보를 포함하도록 개선되었습니다. 이제 `NavController`를 이용한 내비게이션과 `WeatherDetailViewModel`을 통한 데이터 연동이 지원됩니다.  
  (`app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt`)
* **날씨 상세 ViewModel**: 현재 날씨 및 시간대별 예보 데이터를 관리하는 `WeatherDetailViewModel`을 추가하였습니다.  
  (`app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailViewModel.kt`)
  
  다음은 새롭게 구현된 상세 화면의 예시입니다:

<img src="https://github.com/user-attachments/assets/3793d026-60da-405e-9ba7-3afe3e8fc6a1" width="500" />


### 데이터 모델
* **새로운 데이터 클래스**: 날씨 및 예보 데이터를 캡슐화하기 위해 `ForecastItem`과 `WeatherDetail` 데이터 클래스를 도입하였습니다.  
  (`app/src/main/java/com/ben/simpleweather/data/ForecastItem.kt`, `app/src/main/java/com/ben/simpleweather/data/WeatherDetail.kt`)  
  [[1]](diffhunk://#diff-e1a8e63e2f226d0891394257e6524cf3ce8d299463c250fcea01530fd082f07bR1-R7) [[2]](diffhunk://#diff-8b1e0768698fb85286c8ce21f95dd2ef6efd17f04c2953fffa10c8eb5d233b23R1-R19)

### UI 개선
* **드로어블 리소스**: 흐림, 비, 바람, 눈 등 다양한 날씨 상태를 나타내는 벡터 드로어블 리소스를 추가하여 UI를 개선하였습니다.  
  (`app/src/main/res/drawable/outline_ac_unit_24.xml`,  
  `app/src/main/res/drawable/outline_air_24.xml`,  
  `app/src/main/res/drawable/outline_cloud_24.xml`,  
  `app/src/main/res/drawable/outline_explore_24.xml`,  
  `app/src/main/res/drawable/outline_grain_24.xml`,  
  `app/src/main/res/drawable/outline_nights_stay_24.xml`)  
  [[1]](diffhunk://#diff-caf9764f4561f04d3f30d93e7735dfb06db57e1e80cebaefb51cc982b921f63eR1-R5) [[2]](diffhunk://#diff-8011460874dfe7bcab4a4aaa4c662d16e37e6dda1a6db22c19138d2fb43c3dc8R1-R5) [[3]](diffhunk://#diff-b12a653e5efd2aaa6b74af30cbf8c18eb01e266c95e25e8f29674fa4303ae9c0R1-R5) [[4]](diffhunk://#diff-94fa0898b6234fdcb3cf056343fd9fe30b27f3b3a46ec0cedcf7b7c44f85cd44R1-R5) [[5]](diffhunk://#diff-45dc96c33974f95211b9f4c605b559d9b44cf619a76264d577e78f4bd388702cR1-R5) [[6]](diffhunk://#diff-7384d9667f4173679c6bbe322d59fb9689f5552e844b66fdb982c7f844682c09R1-R5)

### 코드베이스 업데이트
* **내비게이션 개선**: `WeatherDetailScreen`으로 `NavController`를 전달할 수 있도록 `SimpleWeatherNavHost`를 수정하였습니다.  
  (`app/src/main/java/com/ben/simpleweather/SimpleWeatherNavHost.kt`)
* **매니페스트 업데이트**: 날씨 데이터를 가져오기 위한 네트워크 요청을 위해 `AndroidManifest.xml`에 `INTERNET` 권한을 추가하였습니다.  
  (`app/src/main/AndroidManifest.xml`)

이번 변경을 통해 사용자 경험을 향상시키고, 상세한 날씨 정보 제공과 시각적 개선, 구조적 확장 가능성을 확보하였습니다.

closes #4

Commit:
feat: 인터넷 권한 추가 (#4)

네트워크 통신을 위해 `AndroidManifest.xml` 파일에 `INTERNET` 권한을 추가했습니다.

Commit:
remove: 불필요한 공백 제거 (#4)

app/build.gradle.kts 파일에서 Coil 의존성 선언 후 불필요한 공백 라인을 제거했습니다.

Commit:
feat: 아이콘 리소스 추가 (#4)

애플리케이션에 사용될 다양한 아이콘들을 추가했습니다.
추가된 아이콘들은 다음과 같습니다:

- 에어컨 (ac_unit)
- 바람 (air)
- 구름 (cloud)
- 나침반 (explore)
- 알갱이 (grain)
- 달 (nights_stay)
- 비 (rainy)
- 가시성 (visibility)
- 물방울 (water_drop)
- 황혼 (wb_twilight)

Commit:
feat: 날씨 상세 정보 관련 문자열 리소스 추가 (#4)

날씨 상세 화면에 필요한 다음 문자열 리소스들을 추가했습니다:
- 뒤로가기
- 시간대별 예보
- 상세 정보
- 강수 확률
- 습도
- 풍속
- 가시거리
- 구름량
- 풍향
- 일출
- 일몰
- 강수량 (3시간)
- 적설량 (3시간)

위 문자열들은 기본(영어) 및 한국어 리소스에 모두 추가되었습니다.

Commit:
feat: 일기 예보 아이템 데이터 클래스 추가 (#4)

일기 예보 정보를 담는 `ForecastItem` 데이터 클래스를 추가했습니다. 이 클래스는 시간, 온도, 아이콘 코드를 필드로 가집니다.

Commit:
feat: 날씨 상세 화면 구현 및 ViewModel 연결 (#4)

날씨 상세 정보를 표시하는 화면을 구현하고, ViewModel과 연결하여 더미 데이터를 표시하도록 구성했습니다.

- `WeatherDetailScreen.kt`: 상세 화면 UI 구성, NavController를 통해 뒤로가기 기능 추가
- `WeatherDetailViewModel.kt`: 상세 화면에서 사용할 ViewModel 생성 및 더미 데이터 로드 로직 추가
- `WeatherDetail.kt`: 상세 날씨 정보 데이터 클래스 정의
- `SimpleWeatherNavHost.kt`: 상세 화면으로 이동 시 NavController 전달하도록 수정

Commit:
refactor: `WeatherDetailViewModel`에서 `cityName` 파라미터 제거 (#4)

`WeatherDetailViewModel`의 `loadWeather` 함수에서 불필요한 `cityName` 파라미터를 제거했습니다.

이 변경으로 인해 `WeatherDetailScreen`에서도 `loadWeather` 함수 호출 시 `cityName` 전달 로직이 제거되었습니다.

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
+        if (weather != null) {
+            WeatherDetailContent(
+                modifier = Modifier.padding(innerPadding),
+                weather = weather!!,
```
이 부분 지적하려고 했는데 코파일럿이 suggest 를 했군요. 아래와 같이 하는 게 좋을 거 같네요.

Benjamin8282 replied:
개선하도록 하겠습니다

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
+                weather = weather!!,
+                forecast = forecast
+            )
+        } else {
```
만일 통신단에서 에러가 나서 weather 가 non-null 로 떨어지지 않는다면 어떻게 될까요? 그러면 indicator 가 계속 나타날 거 같네요.

Benjamin8282 replied:
해결을 위해 weather 대신 WeatherUiState를 사용해 로딩, 성공, 실패 상태를 명시적으로 구분하고, 각 상태에 따라 UI가 분기되도록 리팩토링했습니다.

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
 }
+
+fun degToCompass(deg: Int): String {
+    val directions = listOf(
```
이 부분 const로 뺄 수 있으면 빼도 좋을 거 같긴 합니다

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/data/ForecastItem.kt:
```diff
@@ -0,0 +1,7 @@
+data class ForecastItem(
+    val time: String, // "12PM"
+    val temperature: Int,
+    val iconCode: String
```
끝에 , 하나를 붙여 주세요. trailing comma 라고 불리는 것입니다.

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/data/WeatherDetail.kt:
```diff
@@ -0,0 +1,19 @@
+    val sunrise: Long,              // Unix time
+    val sunset: Long,               // Unix time
+    val rainAmount: Double?,        // mm
+    val snowAmount: Double?         // mm
```
위에서 얘기한 trailing comma 를 붙여 주세요

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
+        if (weather != null) {
+            WeatherDetailContent(
+                modifier = Modifier.padding(innerPadding),
+                weather = weather!!,
```
또한 !! 를 붙이는 건 아주 나쁜 practice 입니다. 물론 여기서는 명시적으로 null check 가 들어가긴 하지만, 만에 하나 null check가 되어 있지 않은 코드에서 !! 를 쓰는 것은 NPE 를 유발할 가능성이 있습니다

Benjamin8282 replied:
사용에 유의하도록 하겠습니다

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
+                )
+                Text(text = weather.description, style = MaterialTheme.typography.bodyMedium)
+                Text(
+                    text = "Feels like ${weather.feelsLike}°C",
```
체감온도를 이야기하는 거 같은데 이 부분도 string resource 로 빼는 게 좋을 거 같아요

Benjamin8282 replied:
분리하도록 하겠습니다

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/features/detail/WeatherDetailScreen.kt:
```diff
@@ -1,12 +1,296 @@
+                    text = "${weather.temperature}°C",
+                    style = MaterialTheme.typography.headlineLarge
+                )
+                Text(text = weather.description, style = MaterialTheme.typography.bodyMedium)
```
여기서 network response 에서 내려오는 영문 응답을 그대로 쓰면 한글 버전에서 통일성을 해칠 거 같아요.
api 응답 값을 참고해서, 변환하는 테이블이 있어야 할 거 같습니다.

Benjamin8282 replied:
api 에선 원하는 언어로 요청이 가능해서, 응답을 한국어로 받는 게 가능합니다.

Commit:
refactor: 상수 및 변수 위치 변경 및 코드 간결화 (#4)

- `WEATHER_ICON_BASE_URL` 및 `directions` 상수를 파일 최상단으로 이동하여 가독성 향상
- `weather` 객체가 null일 경우를 처리하는 `if-else` 문을 `let`과 `run`을 사용하여 간결하게 수정
- `degToCompass` 함수 내 `directions` 리스트를 파일 최상단 상수로 대체

Commit:
feat: 체감 온도 표시 추가 및 문자열 리소스화 (#4)

- 날씨 상세 화면에 체감 온도를 표시하도록 UI를 업데이트했습니다.
- "Feels like" 문자열을 `strings.xml`에 추가하여 다국어 지원 및 코드 가독성을 향상시켰습니다.
- 한국어 환경을 위해 `values-ko/strings.xml`에도 "체감온도" 문자열을 추가했습니다.

Commit:
feat: 날씨 상세 화면 UI 상태 관리 개선 (#4)

- `WeatherUiState` sealed class를 도입하여 로딩, 성공, 에러 상태를 명확하게 관리합니다.
- `WeatherDetailViewModel`에서 `StateFlow`를 사용하여 UI 상태를 노출하고, 더미 데이터 로드 시 성공 또는 에러 상태를 발행합니다.
- `WeatherDetailScreen`에서 `WeatherUiState`를 관찰하여 각 상태에 따라 적절한 UI(로딩 인디케이터, 날씨 정보, 에러 메시지)를 표시하도록 수정합니다.

Commit:
style: 데이터 클래스 필드 끝에 쉼표 추가 (#4)

데이터 클래스 `WeatherDetail`, `ForecastItem`, `WeatherItem`의 마지막 필드 뒤에 쉼표를 추가하여 코드 스타일을 일관성 있게 변경했습니다.

Commit:
refactor: 예외 처리 시 변수명 변경 (#4)

날씨 데이터를 불러오는 데 실패했을 때 발생하는 예외 처리 로직에서 사용되지 않는 예외 변수명을 `_`로 변경하여 코드 가독성을 향상시켰습니다.
